### DISCUSSION TITLE
How to adjust `#[frb(external)]` with Trait bound?

### DISCUSSION THREAD

**Discussion by devgony** [fzyzcjy/flutter_rust_bridge#2101] [Q&A]
First of all, Could you please check if i can use automatic third-party scanning for [this crate](https://crates.io/crates/gluesql)
If not, I want to mirror these methods with Trait bounds
```rs
#[frb(external)]
impl<T: GStore + GStoreMut> Glue<T> {
    pub fn new(storage: T) -> Self {}
    pub async fn plan<Sql: AsRef<str>>(&mut self, sql: Sql) -> Result<Vec<Statement>> {}
    pub async fn execute_stmt(&mut self, statement: &Statement) -> Result<Payload> {}
    pub async fn execute<Sql: AsRef<str>>(&mut self, sql: Sql) -> Result<Vec<Payload>> {}
}
```
But it gets 
```rs
error[E0412]: cannot find type `T` in this scope
  --> pkg/dart/rust/src/api/simple.rs:74:34
   |
74 | impl<T: GStore + GStoreMut> Glue<T> {
   |                                  ^ not found in this scope
```
```rs
error[E0207]: the type parameter `T` is not constrained by the impl trait, self type, or predicates
  --> pkg/dart/rust/src/api/simple.rs:74:6
   |
74 | impl<T: GStore + GStoreMut> Glue<T> {
   |      ^ unconstrained type parameter
```
Doesn't it support `#[frb(external)]` with Trait bound?
Do you know how to solve it?

**Comment by fzyzcjy** [comment]
I briefly checked https://github.com/gluesql/gluesql/blob/main/pkg/javascript/src/lib.rs, the JS api of that Rust crate to see what it looks like.

Looks like it has only several functions instead of many. Then in that case, I guess manually mirror some methods are quicker. (But if you need to mirror a ton of functions, maybe consider scanning the crate automatically.)

> I want to mirror these methods with Trait bounds

Firstly,

```rust
fn plan<Sql: AsRef<str>>(&mut self, sql: Sql)
```

may be

```rust
fn plan(&mut self, sql: String)
```

Secondly, generics are not supported yet (thus the `T` error). Thus I personally suggest checking how https://github.com/gluesql/gluesql/blob/main/pkg/javascript/src/lib.rs is done (it has no generics etc) and maybe follow that way.

This doc may be related as well: https://github.com/fzyzcjy/flutter_rust_bridge/blob/feat/12337/website/docs/guides/third-party/manual/wrappers.md

EDIT: Oh I realize you are 3rd contributor to that repo! Then the answer depends on your needs. For example, originally I thought you are an user, then I suggest creating small API that suffices your needs; but if you are package contributor and want to add official Dart/Flutter support, then that would be another story - e.g. whether you want to expose to users a very rich API. 

EDIT: I see the PR https://github.com/gluesql/gluesql/pull/1503 code. For things like https://github.com/devgony/gluesql/blob/6d059cb1314a97f87b43bb1246a093d6bcb49f8f/pkg/dart/rust/src/api/value.rs#L12, I think it can be automatically scanned. For complex things like the generic `T` you mentioned above, I guess there needs to be some manual work (to avoid code duplication, possibly also done with a Rust macro). One way is to firstly try to scan it automatically (and maybe put `#[frb(ignore)]` on some things), and then manually add other things that cannot be scanned.

I am modifying doc about this, thus please refer to https://github.com/fzyzcjy/flutter_rust_bridge/tree/feat/12337/website/docs/guides/third-party for latest before it is merged.

**Reply by devgony** [reply]
> EDIT: Oh I realize you are 3rd contributor to that repo! Then the answer depends on your needs. For example, originally I thought you are an user, then I suggest creating small API that suffices your needs; but if you are package contributor and want to add official Dart/Flutter support, then that would be another story - e.g. whether you want to expose to users a very rich API.

Even though the JS (or Python) binding of GlueSQL used just simple wrapper, Dart is something different. Cuz it will be the first static typed language we are gonna support. I will make it as a standard of typed language binding.

So as you said, I wanted to expose very rich API. After Dart 3, with the power of switch expression and pattern matching, i thought it would be really cool to use all the types of GlueSQL in Flutter.

> EDIT: I see the PR https://github.com/gluesql/gluesql/pull/1503 code. For things like https://github.com/devgony/gluesql/blob/6d059cb1314a97f87b43bb1246a093d6bcb49f8f/pkg/dart/rust/src/api/value.rs#L12, I think it can be automatically scanned.

At first i thought it would be simple to implement with few converting wrapper like that.
Later, i tried #[frb(mirror(Error)] and it looks like i have to mirror all the types. Cuz the `AST of SQL` and `enum Error` are various in our project and i want flutter users to get rich types well [eg)](https://github.com/gluesql/gluesql/blob/c5f123ad4e095c158c2916b07cde5b90ecdab6d1/pkg/dart/rust/src/api/error.rs#L21)

--- 

Thanks to you, I just saw there is `#[frb(ignore)]`. So i will try automatic scan and ignore first!.

### By the way.. When we automatically scan, Can we set workspace name instead of remote crate name?
Dart binding will be one of GlueSQL's workspace and I want to sync with the latest main branch even though it is not released yet.
```toml
# pkg/dart/Cargo.toml
[dependencies]
gluesql-core.workspace = true
memory-storage.workspace = true
```

**Reply by fzyzcjy** [reply]
I see. Ping me if there are any questions!

Btw, what about converting this discussion to a github issue, since as we discuss further, this UI will quickly fold things except for several comments iirc and thus not very convenient.

> When we automatically scan, Can we set workspace name instead of remote crate name?

Try to specify the crate name, and it uses `cargo metadata` to find that crate, thus it should find the one in you workspace, as long as your cargo.toml dependency points to your workspace.
